### DISCUSSION TITLE
Trouble updating an object

### DISCUSSION THREAD

**Discussion by jonaird** [geldata/gel#3242] [Q&A] [answered]
I'm having a trouble with a query and I just can't figure it out. Here's the problematic query:

```
UPDATE Post filter .id = <uuid>"ac83717e540c11ec9a30cf93e0e71bc1"
         SET { 
           links := (SELECT Post FILTER .id in {<uuid>'ce7c5006-4cf1-11ec-9c35-876348e707a1' })
         };
```

The query doesn't return any errors but it doesn't actually seem to update the Post object. If I query the object for the links property I don't get any results:

```
 select 
....... Post{links:{title}}
.......  filter .id = <uuid>"ac83717e540c11ec9a30cf93e0e71bc1";
{default::Post {links: {}}}
```

The nested query in the update seems to work in that it returns another Post object.
```
SELECT Post FILTER .id in {<uuid>'ce7c5006-4cf1-11ec-9c35-876348e707a1' };
{default::Post {id: ce7c5006-4cf1-11ec-9c35-876348e707a1}}
```

What's going on here?

**Comment by vpetrovykh** [comment, marked as answer]
You are using the same name `Post` twice. You are still within the scope of the top-level meaning of the `Post` when you use it in the `set` clause of `update`. This means that the inner `Post` refers to the object being updated, not an independent one. There are a couple of ways to remedy this:

1) Use `DETACHED`
   ```
   UPDATE Post 
   FILTER .id = <uuid>"ac83717e540c11ec9a30cf93e0e71bc1"
   SET { 
     links := (SELECT DETACHED Post FILTER .id in {<uuid>'ce7c5006-4cf1-11ec-9c35-876348e707a1' })
   };
   ```
2) Use `WITH` to define a new independent way to refer to the set `Post`
   ```
   WITH P := Post
   UPDATE Post 
   FILTER .id = <uuid>"ac83717e540c11ec9a30cf93e0e71bc1"
   SET { 
     links := (SELECT P FILTER .id in {<uuid>'ce7c5006-4cf1-11ec-9c35-876348e707a1' })
   };
   ```
